homeassistant/components/moonraker/config_flow.py

```diff
@@ -0,0 +1,216 @@
+    uuid = None
+    try:
+        uuid = hub.system_info["system_info"]["cpu_info"]["serial_number"]
+    except KeyError:
+        pass
+    return {"title": hub.printer_info.get("hostname"), "unique_id": uuid}
+
+
+class ConfigFlow(config_entries.ConfigFlow, domain=DOMAIN):
+    """Handle a config flow for moonraker."""
+
+    VERSION = 1
+    CONNECTION_CLASS = config_entries.CONN_CLASS_LOCAL_POLL
+    api_key_task = None
+
+    def __init__(self) -> None:
+        """Handle a config flow for OctoPrint."""
+        self.discovery_schema = None
+        self._reauth_entry: config_entries.ConfigEntry | None = None
+
+    @callback
+    def _async_current_hosts(self):
+        """Return a set of hosts."""
+        return {
+            entry.data[CONF_HOST]
+            for entry in self._async_current_entries()
+            if CONF_HOST in entry.data
+        }
+
+    async def async_step_user(
+        self, user_input: dict[str, Any] | None = None
+    ) -> FlowResult:
+        """Handle the initial step."""
+        if user_input is None:
+            data = self.discovery_schema or _schema_with_defaults()
+            return self.async_show_form(step_id="user", data_schema=data)
+
+        if user_input[CONF_HOST]:
+            if (
+                not self._reauth_entry
+                and user_input[CONF_HOST] in self._async_current_hosts()
+            ):
+                return self.async_abort(reason="already_configured")
+
+            errors = {}
+            try:
+                info = await validate_input(self.hass, user_input)
+            except CannotConnect:
+                errors["base"] = "cannot_connect"
+            except InvalidAuth:
+                errors["base"] = "invalid_api_key"
+            except Exception:  # pylint: disable=broad-except
+                _LOGGER.exception("Unexpected exception")
+                errors["base"] = "unknown"
+            else:
+                if self._reauth_entry:
+                    self.hass.config_entries.async_update_entry(
+                        self._reauth_entry,
+                        data=user_input,
+                    )
+                    return self.async_abort(reason="reauth_successful")
+                await self.async_set_unique_id(info.get("unique_id"))
+                self._abort_if_unique_id_configured()
+                return self.async_create_entry(title=info["title"], data=user_input)
+
+        return self.async_show_form(
+            step_id="user",
+            data_schema=_schema_with_defaults(
+                user_input[CONF_HOST],
+                user_input[CONF_PORT],
+                user_input[CONF_SSL],
+                user_input[CONF_API_KEY],
+            ),
+            errors=errors,
+        )
+
+    async def async_step_zeroconf(
+        self, discovery_info: zeroconf.ZeroconfServiceInfo
+    ) -> FlowResult:
+        """Handle discovery flow."""
+        local_name = discovery_info[zeroconf.ATTR_HOSTNAME][:-1]
```

Comment:
Use dataclass properties (see #60561):
```suggestion
        local_name = discovery_info.hostname[:-1]
```